pennylane_lightning/src/gates/GateUtil.cpp

```diff
@@ -73,18 +74,51 @@ auto ValueForKernelHelper(
                                     ValueClass>(kernel);
     }
 }
+
+template <class GateImplementation>
+constexpr auto implementedGatesIterPair =
+    std::pair{&(*std::begin(GateImplementation::implemented_gates)),
+              &(*std::begin(GateImplementation::implemented_gates)) +
+                  GateImplementation::implemented_gates.size()};
+
+template <class TypeList, size_t... Is>
+constexpr auto implementedGatesItersHelper(
+    [[maybe_unused]] std::index_sequence<Is...> indices) {
+    return std::array{
+        implementedGatesIterPair<Util::getNthType<TypeList, Is>>...};
+}
+
+constexpr auto implementedGatesIters() {
+    constexpr auto size = Util::length<AvailableKernels>();
+    return implementedGatesItersHelper<AvailableKernels>(
+        std::make_index_sequence<size>());
+}
+
+template <class TypeList, size_t... Is>
+constexpr auto
+kernelIndicesHelper([[maybe_unused]] std::index_sequence<Is...> indices) {
+    return std::array{
+        std::pair{Util::getNth<TypeList, Is>::Type::kernel_id, Is}...};
+}
+/**
+ * @brief Get the position of the given kernel in AvailabeKernels
+ */
+auto kernelIndices(Gates::KernelType kernel) {
+    constexpr static auto size = Util::length<AvailableKernels>();
+    constexpr static auto kernelIndices =
+        kernelIndicesHelper<AvailableKernels>(std::make_index_sequence<size>());
+    return Util::lookup(kernelIndices, kernel);
+}
```

Comment:
It's a little unclear to me what is happening here. How are these functions used?

Reply:
I will anyway add some docstring for usage.